lib/graphql/schema/field.rb

```diff
@@ -615,59 +614,73 @@ def resolve(object, args, query_ctx)
         query_ctx.schema.after_lazy(self.authorized?(application_object, args, query_ctx)) do |is_authorized|
           if is_authorized
             with_extensions(object, args, query_ctx) do |obj, ruby_kwargs|
-              method_args = ruby_kwargs
+                method_args = ruby_kwargs
               if @resolver_class
                 if obj.is_a?(GraphQL::Schema::Object)
                   obj = obj.object
                 end
                 obj = @resolver_class.new(object: obj, context: query_ctx, field: self)
               end
 
-              # Find a way to resolve this field, checking:
-              #
-              # - A method on the type instance;
-              # - Hash keys, if the wrapped object is a hash;
-              # - A method on the wrapped object;
-              # - Or, raise not implemented.
-              #
-              if obj.respond_to?(resolver_method)
-                method_to_call = resolver_method
-                method_receiver = obj
-                # Call the method with kwargs, if there are any
-                if ruby_kwargs.any?
-                  obj.public_send(resolver_method, **ruby_kwargs)
-                else
-                  obj.public_send(resolver_method)
-                end
-              elsif obj.object.is_a?(Hash)
-                inner_object = obj.object
+              # Resolver Priority (the order in which it will try to resolve a field)
+              # 1 Hash
+              # -> 1.1 Dig Keys
+              # -> 1.2 Hash Key
+              # -> 1.3 Object Method (as a key)
+              # -> 1.4 Resolver Method (as a key)
+              # -> 1.5 Field Name (as a key) - through @resolver_method
+              # 2 Object
+              # -> 2.1 Hash Key
+              # -> 2.2 Object Method
+              # -> 2.3 Resolver Method
+              # -> 2.4 Field Name (as a key) - through @resolver_method
+              # 3 Resolver
+              # -> 2.1 Hash Key
+              # -> 2.2 Object Method
+              # -> 2.3 Resolver Method
+              # -> 2.4 Field Name (as a key) - through @resolver_method
+
+              inner_object = obj.object
+              method_or_key = @hash_key&.to_sym || @object_method&.to_sym || @resolver_method&.to_sym
+
+              if inner_object.is_a?(Hash)
                 if @dig_keys
                   inner_object.dig(*@dig_keys)
-                elsif inner_object.key?(@method_sym)
-                  inner_object[@method_sym]
+                elsif inner_object.key?(method_or_key)
+                  inner_object[method_or_key]
+                elsif inner_object.key?(method_or_key.to_s)
+                  inner_object[method_or_key.to_s]
```

Comment:
It looks like this will create new strings every time the field is resolved. For big responses, this will add up, performance-wise. The previous implementation didn't create any new strings while running -- can we retain that advantage here?

Reply:
4ea87cdf7